### DISCUSSION TITLE
How do I make tree-sitter deal with context-related language

### DISCUSSION THREAD

**Discussion by xis19** [tree-sitter/tree-sitter#3311] [Q&A]
I am working on highlighting syntax for a form-like language, e.g.

```
declare form1 form5 form3

form1-content

form5-content

form3-content

end
```
where form1, 3 and 5 have different syntaxes. It is also possible that the user uses form2 or form4, etc.. And  I am wondering how do I configure tree-sitter for this.

**Comment by wldmr** [comment]
There's not a lot to go on with this question. I general, this seems like a job for the `choice()` [DSL function](https://tree-sitter.github.io/tree-sitter/creating-parsers#the-grammar-dsl).

Can you give an example of your current grammar, plus some specific input (your example is rather vague), and the errors you're getting?

**Reply by xis19** [reply]
My language is Gaussian, which is defined in https://gaussian.com/input/

In the route sections, user will give certain keywords and the define the list of the following sections. Different keywords will cause different combinations of sections and each section may have its own syntax, e.g. XYZ/ZMat or basis set definition or something else.

**Comment by nicolas-graves** [comment]
I have a similar issue, which definitely fits the name of this discussion. I was in the process of writing a tree-sitter grammar a few weeks ago and I stumbled on a case I couldn't handle easily. The language is scilab and it has a way to [convert functions into CLI-style calls](https://help.scilab.org/docs/2024.0.0/en_US/functions.html) : 

> 
> When a function has no left hand side argument and is called only with character string arguments, the calling syntax may be simplified:
> ```
> fun('a','toto','a string')
> ```
> is equivalent to:
> ```
> fun a toto 'a string'
> ```


Due to this behaviour, I'm unable to properly parse this example : 
```
A + 2
```
We generally want to parse this as an addition of identifier and number, but since functions have precedence over addition here, `A` can well be a function that takes "+" and "2" as arguments (tested). Moreover, A can be defined in another file so I'm not even sure we can try to do some smart thing in the file. 

I'm not well-versed into LR grammars, but is it still possible to parse scilab with tree-sitter or does this property make it impossible? (I can still parse a subset without the properly and advise against using it though as a warning).
